src/policy/compressor/forwarding.rs

```diff
@@ -0,0 +1,165 @@
+use crate::policy::compressor::GC_MARK_BIT_MASK;
+use crate::util::constants::MIN_OBJECT_SIZE;
+use crate::util::heap::MonotonePageResource;
+use crate::util::metadata::side_metadata::spec_defs::{COMPRESSOR_MARK, COMPRESSOR_OFFSET_VECTOR};
+use crate::util::metadata::side_metadata::SideMetadataSpec;
+use crate::util::{Address, ObjectReference};
+use crate::vm::object_model::ObjectModel;
+use crate::vm::VMBinding;
+use atomic::Ordering;
+use std::marker::PhantomData;
+use std::sync::atomic::AtomicBool;
+
+/// A finite-state machine which processes the positions of mark bits,
+/// and accumulates the size of live data that it has seen.
+///
+/// The Compressor caches the state of the transducer at the start of
+/// each block by serialising the state using [`Transducer::encode`];
+/// the state can then be deserialised using [`Transducer::decode`].
+#[derive(Debug)]
+struct Transducer {
+    live: usize,
+    last_bit_seen: Address,
+    in_object: bool,
+}
+impl Transducer {
+    pub fn new() -> Self {
+        Self {
+            live: 0,
+            last_bit_seen: Address::ZERO,
+            in_object: false,
+        }
+    }
+    pub fn step(&mut self, address: Address) {
+        if self.in_object {
+            self.live += address - self.last_bit_seen + MIN_OBJECT_SIZE;
```

Comment:
Why do we still need `MIN_OBJECT_SIZE` here?

Reply:
> The size of an object is the difference between its end and its start; the last bit is the previous word from the end, so the distance between the two mark bits is off by one word. I'll comment this too.

So if it is currently 'in_object', the address would be the 'end address', and the 'last_bit_seen' would be the start of the object. The diff plus 'MIN_OBJECT_SIZE' (or whatever you use to calculate the end address) would be the object size. Right?

You can make this more clear with code. It is more verbose, but clearer.
```rust
if self.in_object {
  let obj_end = address;
  let obj_start = self.last_bit_seen;
  let obj_size = obj_end - obj_start + X;
  self.live += obj_size;
}
```

Reply:
> Right?

Right. I like this idea.

Reply:
I've done this in 1ef15cf